src/base_alloc/base_alloc_linear.c

```diff
@@ -0,0 +1,163 @@
+    os_mutex_t *lock;
+    char *data_ptr;
+    size_t size_left;
+#ifndef NDEBUG
+    size_t n_pools;
+#endif /* NDEBUG */
+};
+
+// the main pool of the linear base allocator (there is only one such pool)
+struct umf_ba_linear_pool_t {
+    // address of the beginning of the next pool (a list of allocated pools to be freed in umf_ba_linear_destroy())
+    umf_ba_next_linear_pool_t *next_pool;
+
+    // metadata is set and used only in the main (the first) pool
+    struct umf_ba_main_linear_pool_meta_t metadata;
+
+    // data area of the main pool (the first one) starts here
+    char data[];
+};
+
+// the "next" pools of the linear base allocator (pools allocated later, when we run out of the memory of the main pool)
+struct umf_ba_next_linear_pool_t {
+    // address of the beginning of the next pool (a list of allocated pools to be freed in umf_ba_linear_destroy())
+    umf_ba_next_linear_pool_t *next_pool;
+
+    // data area of all pools except of the main (the first one) starts here
+    char data[];
+};
+
+#ifndef NDEBUG
+static void ba_debug_checks(umf_ba_linear_pool_t *pool) {
+    // count pools
+    size_t n_pools = 1;
+    umf_ba_next_linear_pool_t *next_pool = pool->next_pool;
+    while (next_pool) {
+        n_pools++;
+        next_pool = next_pool->next_pool;
+    }
+    assert(n_pools == pool->metadata.n_pools);
+}
+#endif /* NDEBUG */
+
+umf_ba_linear_pool_t *umf_ba_linear_create(size_t pool_size) {
+    size_t mutex_size = align_size(util_mutex_get_size(), MEMORY_ALIGNMENT);
+    size_t metadata_size = sizeof(struct umf_ba_main_linear_pool_meta_t);
+    pool_size = pool_size + metadata_size + mutex_size;
+    if (pool_size < MINIMUM_LINEAR_POOL_SIZE) {
+        pool_size = MINIMUM_LINEAR_POOL_SIZE;
+    }
+
+    pool_size = align_size(pool_size, ba_os_get_page_size());
+
+    umf_ba_linear_pool_t *pool = (umf_ba_linear_pool_t *)ba_os_alloc(pool_size);
+    if (!pool) {
+        return NULL;
+    }
+
+    void *data_ptr = &pool->data;
+    size_t size_left = pool_size - offsetof(umf_ba_linear_pool_t, data);
+
+    align_ptr_size_left(&data_ptr, &size_left, MEMORY_ALIGNMENT);
+
+    pool->metadata.pool_size = pool_size;
+    pool->metadata.data_ptr = data_ptr;
+    pool->metadata.size_left = size_left;
+    pool->next_pool = NULL; // this is the only pool now
+#ifndef NDEBUG
+    pool->metadata.n_pools = 1;
+#endif /* NDEBUG */
+
+    // allocate and init lock
+    pool->metadata.lock = util_mutex_init(pool->metadata.data_ptr);
+    if (!pool->metadata.lock) {
+        ba_os_free(pool, pool_size);
+        return NULL;
+    }
+
+    pool->metadata.data_ptr += mutex_size;  // lock is here
+    pool->metadata.size_left -= mutex_size; // for lock
+
+    return pool;
+}
+
+void *umf_ba_linear_alloc(umf_ba_linear_pool_t *pool, size_t size) {
+    size_t aligned_size = align_size(size, MEMORY_ALIGNMENT);
+    util_mutex_lock(pool->metadata.lock);
+    if (pool->metadata.size_left < aligned_size) {
```

Comment:
since linear alloc is used only internally we should never run out of space - just preallocate more in the initialization

Reply:
Done